engine/packages/pegboard/src/workflows/serverless/pool.rs

```diff
@@ -0,0 +1,195 @@
+use futures_util::FutureExt;
+use gas::{db::WorkflowData, prelude::*};
+use rivet_types::{keys, runner_configs::RunnerConfigKind};
+
+use super::runner;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct Input {
+	pub namespace_id: Id,
+	pub runner_name: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, Default)]
+struct LifecycleState {
+	runners: Vec<Id>,
+}
+
+#[workflow]
+pub async fn pegboard_serverless_pool(ctx: &mut WorkflowCtx, input: &Input) -> Result<()> {
+	ctx.loope(LifecycleState::default(), |ctx, state| {
+		let input = input.clone();
+		async move {
+			// 1. Remove completed connections
+			let completed_runners = ctx
+				.activity(GetCompletedInput {
+					runners: state.runners.clone(),
+				})
+				.await?;
+
+			state.runners.retain(|r| !completed_runners.contains(r));
+
+			// 2. Get desired count -> drain and start counts
+			let ReadDesiredOutput::Desired(desired_count) = ctx
+				.activity(ReadDesiredInput {
+					namespace_id: input.namespace_id,
+					runner_name: input.runner_name.clone(),
+				})
+				.await?
+			else {
+				return Ok(Loop::Break(()));
+			};
+
+			let drain_count = state.runners.len().saturating_sub(desired_count);
+			let start_count = desired_count.saturating_sub(state.runners.len());
+
+			// 3. Drain old runners
+			if drain_count != 0 {
+				// TODO: Implement smart logic of draining runners with the lowest allocated actors
+				let draining_runners = state.runners.iter().take(drain_count).collect::<Vec<_>>();
+
+				for wf_id in draining_runners {
+					ctx.signal(runner::Drain {})
+						.to_workflow_id(*wf_id)
+						.send()
+						.await?;
+				}
+			}
+
+			// 4. Dispatch new runner workflows
+			if start_count != 0 {
+				for _ in 0..start_count {
+					let wf_id = ctx
+						.workflow(runner::Input {
+							pool_wf_id: ctx.workflow_id(),
+							namespace_id: input.namespace_id,
+							runner_name: input.runner_name.clone(),
+						})
+						.tag("namespace_id", input.namespace_id)
+						.tag("runner_name", input.runner_name.clone())
+						.dispatch()
+						.await?;
+
+					state.runners.push(wf_id);
+				}
+			}
+
+			// Wait for Bump or runner update signals until we tick again
+			match ctx.listen::<Main>().await? {
+				Main::RunnerDrainStarted(sig) => {
+					state.runners.retain(|wf_id| *wf_id != sig.runner_wf_id);
+				}
+				Main::Bump(_) => {}
+			}
+
+			Ok(Loop::Continue)
+		}
+		.boxed()
+	})
+	.await?;
+
+	Ok(())
+}
+
+#[derive(Debug, Serialize, Deserialize, Hash)]
+struct GetCompletedInput {
+	runners: Vec<Id>,
+}
+
+#[activity(GetCompleted)]
+async fn get_completed(ctx: &ActivityCtx, input: &GetCompletedInput) -> Result<Vec<Id>> {
+	Ok(ctx
```

Comment:
this is not going to scale well. thoughts on having runners send sig to pool when complete so we can use batch listen?